// appium/build/lib/extension/extension-config.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.INSTALL_TYPE_NPM = exports.INSTALL_TYPE_LOCAL = exports.INSTALL_TYPE_GITHUB = exports.INSTALL_TYPE_GIT = exports.INSTALL_TYPES = exports.ExtensionConfig = void 0;

require("source-map-support/register");

var _lodash = _interopRequireDefault(require("lodash"));

var _bluebird = _interopRequireDefault(require("bluebird"));

var _path = _interopRequireDefault(require("path"));

var _resolveFrom = _interopRequireDefault(require("resolve-from"));

var _semver = require("semver");

var _support = require("@appium/support");

var _extension = require("../cli/extension");

var _config = require("../config");

var _logger = _interopRequireDefault(require("../logger"));

var _schema = require("../schema/schema");

const INSTALL_TYPE_NPM = 'npm';
exports.INSTALL_TYPE_NPM = INSTALL_TYPE_NPM;
const INSTALL_TYPE_LOCAL = 'local';
exports.INSTALL_TYPE_LOCAL = INSTALL_TYPE_LOCAL;
const INSTALL_TYPE_GITHUB = 'github';
exports.INSTALL_TYPE_GITHUB = INSTALL_TYPE_GITHUB;
const INSTALL_TYPE_GIT = 'git';
exports.INSTALL_TYPE_GIT = INSTALL_TYPE_GIT;
const INSTALL_TYPES = new Set([INSTALL_TYPE_GIT, INSTALL_TYPE_GITHUB, INSTALL_TYPE_LOCAL, INSTALL_TYPE_NPM]);
exports.INSTALL_TYPES = INSTALL_TYPES;

class ExtensionConfig {
  extensionType;
  configKey;
  installedExtensions;
  log;
  manifest;
  _listDataCache;

  constructor(extensionType, manifest) {
    this.extensionType = extensionType;
    this.configKey = `${extensionType}s`;
    this.installedExtensions = manifest.getExtensionData(extensionType);
    this.manifest = manifest;
  }

  get manifestPath() {
    return this.manifest.manifestPath;
  }

  get appiumHome() {
    return this.manifest.appiumHome;
  }

  getProblems(extName, extManifest) {
    return [...this.getGenericConfigProblems(extManifest, extName), ...this.getConfigProblems(extManifest, extName), ...this.getSchemaProblems(extManifest, extName)];
  }

  async getWarnings(extName, extManifest) {
    const [genericConfigWarnings, configWarnings] = await _bluebird.default.all([this.getGenericConfigWarnings(extManifest, extName), this.getConfigWarnings(extManifest, extName)]);
    return [...genericConfigWarnings, ...configWarnings];
  }

  async getConfigWarnings(extManifest, extName) {
    return [];
  }

  getValidationResultSummaries(errorMap = new Map(), warningMap = new Map()) {
    const errorSummaries = [];

    for (const [extName, problems] of errorMap.entries()) {
      if (_lodash.default.isEmpty(problems)) {
        continue;
      }

      errorSummaries.push(`${this.extensionType} "${extName}" had ${_support.util.pluralize('error', problems.length)} and will not be available:`);

      for (const problem of problems) {
        errorSummaries.push(`  - ${problem.err} (Actual value: ` + `${JSON.stringify(problem.val)})`);
      }
    }

    const warningSummaries = [];

    for (const [extName, warnings] of warningMap.entries()) {
      if (_lodash.default.isEmpty(warnings)) {
        continue;
      }

      const extTypeText = _lodash.default.capitalize(this.extensionType);

      const problemEnumerationText = _support.util.pluralize('potential problem', warnings.length, true);

      warningSummaries.push(`${extTypeText} "${extName}" has ${problemEnumerationText}: `);

      for (const warning of warnings) {
        warningSummaries.push(`  - ${warning}`);
      }
    }

    return {
      errorSummaries,
      warningSummaries
    };
  }

  async _validate(exts) {
    const errorMap = new Map();
    const warningMap = new Map();

    for (const [extName, extManifest] of _lodash.default.toPairs(exts)) {
      const [errors, warnings] = await _bluebird.default.all([this.getProblems(extName, extManifest), this.getWarnings(extName, extManifest)]);

      if (errors.length) {
        delete exts[extName];
      }

      errorMap.set(extName, errors);
      warningMap.set(extName, warnings);
    }

    const {
      errorSummaries,
      warningSummaries
    } = this.getValidationResultSummaries(errorMap, warningMap);

    if (!_lodash.default.isEmpty(errorSummaries)) {
      _logger.default.error(`Appium encountered ${_support.util.pluralize('error', errorSummaries.length, true)} while validating ${this.configKey} found in manifest ${this.manifestPath}`);

      for (const summary of errorSummaries) {
        _logger.default.error(summary);
      }
    } else {
      if (!_lodash.default.isEmpty(warningSummaries)) {
        _logger.default.warn(`Appium encountered ${_support.util.pluralize('warning', warningSummaries.length, true)} while validating ${this.configKey} found in manifest ${this.manifestPath}`);

        for (const summary of warningSummaries) {
          _logger.default.warn(summary);
        }
      }
    }

    return exts;
  }

  async getListData() {
    if (this._listDataCache) {
      return this._listDataCache;
    }

    const CommandClass = _extension.commandClasses[this.extensionType];
    const cmd = new CommandClass({
      config: this,
      json: true
    });
    const listData = await cmd.list({
      showInstalled: true,
      showUpdates: true
    });
    this._listDataCache = listData;
    return listData;
  }

  async getGenericConfigWarnings(extManifest, extName) {
    const {
      appiumVersion,
      installSpec,
      installType,
      pkgName
    } = extManifest;
    const warnings = [];
    const invalidFields = [];

    if (!_lodash.default.isString(installSpec)) {
      invalidFields.push('installSpec');
    }

    if (!INSTALL_TYPES.has(installType)) {
      invalidFields.push('installType');
    }

    const extTypeText = _lodash.default.capitalize(this.extensionType);

    if (invalidFields.length) {
      const invalidFieldsEnumerationText = _support.util.pluralize('invalid or missing field', invalidFields.length, true);

      const invalidFieldsText = invalidFields.map(field => `"${field}"`).join(', ');
      warnings.push(`${extTypeText} "${extName}" (package \`${pkgName}\`) has ${invalidFieldsEnumerationText} (${invalidFieldsText}) in \`extensions.yaml\`; this may cause upgrades done via the \`appium\` CLI tool to fail. Please reinstall with \`appium ${this.extensionType} uninstall ${extName}\` and \`appium ${this.extensionType} install ${extName}\` to attempt a fix.`);
    }

    const createPeerWarning = reason => `${extTypeText} "${extName}" (package \`${pkgName}\`) may be incompatible with the current version of Appium (v${_config.APPIUM_VER}) due to ${reason}`;

    if (_lodash.default.isString(appiumVersion) && !(0, _semver.satisfies)(_config.APPIUM_VER, appiumVersion)) {
      const listData = await this.getListData();
      const extListData = listData[extName];

      if (extListData !== null && extListData !== void 0 && extListData.installed) {
        const {
          updateVersion,
          upToDate
        } = extListData;

        if (!upToDate) {
          warnings.push(createPeerWarning(`its peer dependency on older Appium v${appiumVersion}. Please upgrade \`${pkgName}\` to v${updateVersion} or newer.`));
        } else {
          warnings.push(createPeerWarning(`its peer dependency on older Appium v${appiumVersion}. Please ask the developer of \`${pkgName}\` to update the peer dependency on Appium to v${_config.APPIUM_VER}.`));
        }
      }
    } else if (!_lodash.default.isString(appiumVersion)) {
      const listData = await this.getListData();
      const extListData = listData[extName];

      if (!(extListData !== null && extListData !== void 0 && extListData.upToDate) && extListData !== null && extListData !== void 0 && extListData.updateVersion) {
        warnings.push(createPeerWarning(`an invalid or missing peer dependency on Appium. A newer version of \`${pkgName}\` is available; please attempt to upgrade "${extName}" to v${extListData.updateVersion} or newer.`));
      } else {
        warnings.push(createPeerWarning(`an invalid or missing peer dependency on Appium. Please ask the developer of \`${pkgName}\` to add a peer dependency on \`^appium@${_config.APPIUM_VER}\`.`));
      }
    }

    return warnings;
  }

  getSchemaProblems(extManifest, extName) {
    const problems = [];
    const {
      schema: argSchemaPath
    } = extManifest;

    if (ExtensionConfig.extDataHasSchema(extManifest)) {
      if (_lodash.default.isString(argSchemaPath)) {
        if ((0, _schema.isAllowedSchemaFileExtension)(argSchemaPath)) {
          try {
            this.readExtensionSchema(extName, extManifest);
          } catch (err) {
            problems.push({
              err: `Unable to register schema at path ${argSchemaPath}; ${err.message}`,
              val: argSchemaPath
            });
          }
        } else {
          problems.push({
            err: `Schema file has unsupported extension. Allowed: ${[..._schema.ALLOWED_SCHEMA_EXTENSIONS].join(', ')}`,
            val: argSchemaPath
          });
        }
      } else if (_lodash.default.isPlainObject(argSchemaPath)) {
        try {
          this.readExtensionSchema(extName, extManifest);
        } catch (err) {
          problems.push({
            err: `Unable to register embedded schema; ${err.message}`,
            val: argSchemaPath
          });
        }
      } else {
        problems.push({
          err: 'Incorrectly formatted schema field; must be a path to a schema file or a schema object.',
          val: argSchemaPath
        });
      }
    }

    return problems;
  }

  getGenericConfigProblems(extManifest, extName) {
    const {
      version,
      pkgName,
      mainClass
    } = extManifest;
    const problems = [];

    if (!_lodash.default.isString(version)) {
      problems.push({
        err: `Invalid or missing \`version\` field in my \`package.json\` and/or \`extensions.yaml\` (must be a string)`,
        val: version
      });
    }

    if (!_lodash.default.isString(pkgName)) {
      problems.push({
        err: `Invalid or missing \`name\` field in my \`package.json\` and/or \`extensions.yaml\` (must be a string)`,
        val: pkgName
      });
    }

    if (!_lodash.default.isString(mainClass)) {
      problems.push({
        err: `Invalid or missing \`appium.mainClass\` field in my \`package.json\` and/or \`mainClass\` field in \`extensions.yaml\` (must be a string)`,
        val: mainClass
      });
    }

    return problems;
  }

  getConfigProblems(extManifest, extName) {
    return [];
  }

  async addExtension(extName, extManifest, {
    write = true
  } = {}) {
    this.manifest.addExtension(this.extensionType, extName, extManifest);

    if (write) {
      await this.manifest.write();
    }
  }

  async updateExtension(extName, extManifest, {
    write = true
  } = {}) {
    this.installedExtensions[extName] = { ...this.installedExtensions[extName],
      ...extManifest
    };

    if (write) {
      await this.manifest.write();
    }
  }

  async removeExtension(extName, {
    write = true
  } = {}) {
    delete this.installedExtensions[extName];

    if (write) {
      await this.manifest.write();
    }
  }

  print(activeNames) {
    if (_lodash.default.isEmpty(this.installedExtensions)) {
      _logger.default.info(`No ${this.configKey} have been installed in ${this.appiumHome}. Use the "appium ${this.extensionType}" ` + 'command to install the one(s) you want to use.');

      return;
    }

    _logger.default.info(`Available ${this.configKey}:`);

    for (const [extName, extManifest] of _lodash.default.toPairs(this.installedExtensions)) {
      _logger.default.info(`  - ${this.extensionDesc(extName, extManifest)}`);
    }
  }

  extensionDesc(extName, extManifest) {
    throw new Error('This must be implemented in a subclass');
  }

  getInstallPath(extName) {
    return _path.default.join(this.appiumHome, 'node_modules', this.installedExtensions[extName].pkgName);
  }

  require(extName) {
    const {
      mainClass
    } = this.installedExtensions[extName];
    const reqPath = this.getInstallPath(extName);
    let reqResolved;

    try {
      reqResolved = require.resolve(reqPath);
    } catch (err) {
      throw new ReferenceError(`Could not find a ${this.extensionType} installed at ${reqPath}`);
    }

    if (process.env.APPIUM_RELOAD_EXTENSIONS && require.cache[reqResolved]) {
      _logger.default.debug(`Removing ${reqResolved} from require cache`);

      delete require.cache[reqResolved];
    }

    _logger.default.debug(`Requiring ${this.extensionType} at ${reqPath}`);

    const MainClass = require(reqPath)[mainClass];

    if (!MainClass) {
      throw new ReferenceError(`Could not find a class named "${mainClass}" exported by ${this.extensionType} "${extName}"`);
    }

    return MainClass;
  }

  isInstalled(extName) {
    return _lodash.default.includes(Object.keys(this.installedExtensions), extName);
  }

  static _readExtensionSchema(appiumHome, extType, extName, extManifest) {
    const {
      pkgName,
      schema: argSchemaPath
    } = extManifest;

    if (!argSchemaPath) {
      throw new TypeError(`No \`schema\` property found in config for ${extType} ${pkgName} -- why is this function being called?`);
    }

    let moduleObject;

    if (_lodash.default.isString(argSchemaPath)) {
      const schemaPath = (0, _resolveFrom.default)(appiumHome, _path.default.join(pkgName, argSchemaPath));
      moduleObject = require(schemaPath);
    } else {
      moduleObject = argSchemaPath;
    }

    const schema = moduleObject.__esModule ? moduleObject.default : moduleObject;
    (0, _schema.registerSchema)(extType, extName, schema);
    return schema;
  }

  static extDataHasSchema(extManifest) {
    return _lodash.default.isString(extManifest === null || extManifest === void 0 ? void 0 : extManifest.schema) || _lodash.default.isObject(extManifest === null || extManifest === void 0 ? void 0 : extManifest.schema);
  }

  readExtensionSchema(extName, extManifest) {
    return ExtensionConfig._readExtensionSchema(this.appiumHome, this.extensionType, extName, extManifest);
  }

}

exports.ExtensionConfig = ExtensionConfig;
